src/main/java/uk/nhs/digital/nhsconnect/nhais/parse/EdifactParserV2.java

```diff
@@ -0,0 +1,143 @@
+            .map(messageStartEndIndexPair -> wholeInterchangeLines.subList(messageStartEndIndexPair.getLeft(), messageStartEndIndexPair.getRight() + 1))
+            .map(this::parseMessage)
+            .collect(Collectors.toList());
+    }
+
+    private MessageV2 parseMessage(List<String> wholeMessageLines) {
+        var transactionStartIndexes = findAllIndexes(wholeMessageLines, ReferenceTransactionNumber.KEY_QUALIFIER);
+        var transactionEndIndexes = new ArrayList<>(transactionStartIndexes);
+        transactionEndIndexes.remove(0); // there is no tran end so ending index is the beginning of the next one so end is beginning shifted 1 right
+        transactionEndIndexes.add(wholeMessageLines.size() - 1); // to remove message trailer
+
+        var transactionIndexPairs = zipIndexes(transactionStartIndexes, transactionEndIndexes);
+
+        var transactions = transactionIndexPairs.stream()
+            .map(transactionIndexPair -> wholeMessageLines.subList(transactionIndexPair.getLeft(), transactionIndexPair.getRight()))
+            .map(TransactionV2::new)
+            .collect(Collectors.toList());
+
+        var onlyMessageLines = new ArrayList<>(wholeMessageLines.subList(0, transactionStartIndexes.get(0))); // first lines until transaction
+        onlyMessageLines.add(wholeMessageLines.get(wholeMessageLines.size() - 1)); // message trailer
+
+        var message = new MessageV2(onlyMessageLines);
+        message.setTransactions(transactions);
+        transactions.forEach(transaction -> transaction.setMessage(message));
+
+        return message;
+    }
+
+    private boolean areIndexesInOrder(List<Pair<Integer, Integer>> messageIndexPairs) {
+        return Comparators.isInOrder(
+            messageIndexPairs.stream()
+                .map(pair -> List.of(pair.getLeft(), pair.getRight()))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList()),
+            Comparator.naturalOrder());
+    }
+
+    private List<String> extractInterchangeLines(List<String> wholeInterchangeLines) {
+
+        var firstMessageHeaderIndex = findAllIndexes(wholeInterchangeLines, MessageHeader.KEY).get(0);
+        var allMessageTrailerIndexes = findAllIndexes(wholeInterchangeLines, MessageTrailer.KEY);
+        var lastMessageTrailerIndex = allMessageTrailerIndexes.get(allMessageTrailerIndexes.size() - 1);
+
+        return Stream.of(
+            wholeInterchangeLines.subList(0, firstMessageHeaderIndex),
+            wholeInterchangeLines.subList(lastMessageTrailerIndex + 1, wholeInterchangeLines.size()))
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList());
+    }
+
+    private List<Integer> findAllIndexes(List<String> list, String key) {
+        var indexes = new ArrayList<Integer>();
+        for (int i = 0; i < list.size(); i++) {
+            if (list.get(i).startsWith(key)) {
+                indexes.add(i);
+            }
+        }
+        return indexes;
```

Comment:
I don't think it will iterate more than once, as streams use lazy evaluation: https://www.logicbig.com/tutorials/core-java-tutorial/java-util-stream/lazy-evaluation.html

Reply:
but executing `indexOf` does a list lookup which is O(n)